File: python/421. Maximum XOR of Two Numbers in an Array.py
class TrieNode:
    def __init__(self) -> None:
        self.nodes = [None, None]
        self.val = 0

class Trie:
    NUM_BIT = 31
    def __init__(self):
        self.root = TrieNode()
    
    def add_number(self, number):
        cur_node = self.root
        for bit_id in range(Trie.NUM_BIT, -1, -1):
            cur_bit = (number >> bit_id) & 1
            if cur_node.nodes[cur_bit] is None:
                cur_node.nodes[cur_bit] = TrieNode()
            cur_node = cur_node.nodes[cur_bit]
        cur_node.val = number

class Solution:
    def findMaximumXOR(self, nums: List[int]) -> int:
        trie = Trie()
        max_ans = 0
        for num in nums:
            trie.add_number(num)
        for num in nums:
            node = trie.root
            for bit_id in range(Trie.NUM_BIT, -1, -1):
                cur_bit = (num >> bit_id) & 1

                need_val = 1 - cur_bit
                node = node.nodes[need_val] if node.nodes[need_val] else node.nodes[cur_bit]
            max_ans = max(max_ans, num ^ node.val)
        return max_ans


class Solution2:
    def findMaximumXOR(self, nums: List[int]) -> int:
        max_xor = 0
        mask = 0
        
        for i in range(31, -1, -1):
            mask |= (1 << i)
            prefixes = {num & mask for num in nums}
            temp = max_xor | (1 << i)
            if any(temp ^ p in prefixes for p in prefixes):
                max_xor = temp
        
        return max_xor

File: rust/421. Maximum XOR of Two Numbers in an Array.rs
use std::cmp;

struct Solution;

struct TrieNode {
    nodes: [Option<Box<TrieNode>>; 2],
    val: i32,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            nodes: [None, None],
            val: 0,
        }
    }
}

struct Trie {
    root: Box<TrieNode>,
}

impl Trie {
    const NUM_BIT: i32 = 31;
    fn new() -> Self {
        Trie {
            root: Box::new(TrieNode::new()),
        }
    }

    fn add_number(&mut self, number: i32) {
        let mut cur_node = &mut self.root;
        for bit_id in (0..Self::NUM_BIT).rev() {
            let cur_bit = (number >> bit_id) & 1;
            cur_node = cur_node.nodes[cur_bit as usize].get_or_insert(Box::new(TrieNode::new()));
        }
        cur_node.val = number;
    }
}

impl Solution {
    pub fn find_maximum_xor(nums: Vec<i32>) -> i32 {
        let mut trie = Trie::new();
        for num in nums.iter() {
            trie.add_number(*num);
        }

        let mut max_ans = 0;

        for num in nums.iter() {
            let mut node = &trie.root;
            for bit in (0..Trie::NUM_BIT).rev() {
                let cur_bit = (num >> bit) & 1;
                let need_val = 1 - cur_bit;
                node = match &node.nodes[need_val as usize] {
                    Some(n) => n,
                    None => {node.nodes[cur_bit as usize].as_ref().unwrap()},
                };
            }
            max_ans = cmp::max(max_ans, num ^ node.val);
        }

        max_ans
    }
}
fn main() {
    let nums = vec![3, 10, 5, 25, 2];
    assert_eq!(Solution::find_maximum_xor(nums), 28); // The maximum XOR is 5 ^ 27 = 28
}
